src/node_builtins.cc

```diff
@@ -54,49 +53,50 @@ BuiltinLoader::BuiltinLoader() : config_(GetConfig()), has_code_cache_(false) {
 #endif  // NODE_SHARED_BUILTIN_UNDICI_UNDICI_PATH
 }
 
-BuiltinLoader* BuiltinLoader::GetInstance() {
-  return &instance_;
-}
-
 bool BuiltinLoader::Exists(const char* id) {
-  auto& source = GetInstance()->source_;
-  return source.find(id) != source.end();
+  auto source = source_.read();
+  return source->find(id) != source->end();
 }
 
 bool BuiltinLoader::Add(const char* id, const UnionBytes& source) {
-  auto result = GetInstance()->source_.emplace(id, source);
+  builtin_categories_
```

Comment:
`builtin_categories` seems to be something that can just be bundled with the sources as part of a COW member? And I think we can take a `can_be_required` parameter here (true by default for externalized builtins and false by default for main scripts) that can be used to update the categories in the new copy..

Reply:
I’ve tried to implement this but it would mean having to give up the caching/lazy-loading functionality of `BuiltinCategories&`, since otherwise using a shared COW member would mean that writes to the cache would create copies (which they shouldn’t)… so I guess we could do that, and I see that it would make sense to add `can_be_required` as a parameter, but I’m not sure if it’s really worth the complexity?

Reply:
Actually looking at this again, if we don't even intend to reuse the categories across threads, we just do not even need this `builtin_categories_` cache in `BuiltinLoader`? This is only used in tests anyway, we can just move the computation code in `InitializeBuiltinCategories()` to `GetBuiltinCategories()`, compute that two `std::set` on-stack there, and throw them away once they get turned into JS sets, that'd simplify things a lot.